### PR TITLE
CI on Daint via Sarus

### PR DESCRIPTION
**Pull request by haampie**
Adds Dockerfiles to build arbor's tests s.t. they can be run on Daint via Sarus

The CI pipeline has to be stored outside of this repository (https://gitlab.com/cscs-ci/arbor-sim/arbor-ci/-/blob/master/.gitlab-ci.yml) for security reasons, it could be moved to the `arbor` organization on Github if that's more convenient.

Todo:

- [x] A maintainer of the arbor Github project should be added to the `cscs-ci` group on Gitlab, and should mirror this repo at https://gitlab.com/cscs-ci/arbor-sim/arbor via **New Project** > **CI/CD for external repositories** and then pick the `cscs-ci/arbor-sim` namespace.
- [x] Add sensible `srun` commands to the pipeline (https://gitlab.com/cscs-ci/arbor-sim/arbor-ci/-/blob/master/.gitlab-ci.yml#L45-48)
- [x] Install [Bors](https://github.com/bors-ng/bors-ng) for this repo ([installation link](https://github.com/apps/bors/installations/new)), which ensures Gitlab's CI is triggered, and adds other useful features (e.g. authorization for CI)
- [x] Add CI status badge to `README.md`
- [x] Make the arbor-ci repo a submodule of this for convenience
- [x] Make a slurm allocation once to improve CI run times

### CODE REVIEW

File: bors.toml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+status = [
+  "ci/gitlab/%",
+]
+delete_merged_branches = true
```
Comment (bcumming) [inline bors.toml:4]: Does this delete the merged branch from the fork, or does it delete the temporary merge between the PR branch and master?

File: docker/deploy/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      -e /arbor/build/bin/unit-modcc \
+      -e /arbor/build/bin/unit-mpi && \
+    rm -rf /arbor
+
+FROM ubuntu:18.04
+
+# This is the only thing necessary really from nvidia/cuda's ubuntu18.04 runtime image
+ENV NVIDIA_VISIBLE_DEVICES all
+ENV NVIDIA_DRIVER_CAPABILITIES compute,utility
+ENV NVIDIA_REQUIRE_CUDA "cuda>=10.1 brand=tesla,driver>=384,driver<385 brand=tesla,driver>=396,driver<397 brand=tesla,driver>=410,driver<411"
```
Comment (bcumming) [inline docker/deploy/Dockerfile:38]: Why is `NVIDIA_REQUIRE_CUDA` needed?

File: bors.toml (hunk 1 above)
Comment (haampie) [reply bors.toml:4]: I had to check bors's source for this, but it seems this removes branches inside the current repo after they are merged as part of a PR. It does not / cannot delete branches of forks.

File: docker/deploy/Dockerfile (hunk 2 above)
Comment (haampie) [reply docker/deploy/Dockerfile:38]: Normally you would extend NVIDIA's base image, but that's quite big (1GB) as it pulls in the runtime, cublas etc. The way docker and sarus support gpu's is via NVIDIA Container Toolkit, which as a bare minimum just requires those environment variables in the container which are used to check whether the driver (outside of the container) and the container are compatible.

See https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/ubuntu18.04/10.1/base/Dockerfile#L30-33 for reference.

**Comment by haampie** [comment]
> Does this workflow run the unit tests? I can see how they are packaged up, but not where they are executed.

The pipeline is stored outside of this repository to ensure that no external users can open a PR to insert malicious commands. See https://gitlab.com/cscs-ci/arbor-sim/arbor-ci/-/blob/master/.gitlab-ci.yml#L45-48 for the list of tests that are executed. This is part of a separate repo that could be moved to github as well.

FWIW, in DLA-Future I'm trying to automate running tests properly, as they use `ctest` to generate `srun` commands. It would be a bit awkward to remember to add new run commands to the gitlab pipeline each time there's a new tests. It's not completely obvious how to fix this in a safe way, since `ctest` is running inside of the container, whereas the commands it would generate (a list of `srun ... sarus run ... test_a` commands) can only be run from outside of the container.

>  And how is the success/fail communicated to the "caller"?

See https://github.com/eth-cscs/COSMA/pull/30#issuecomment-610601157 for an example. I have not checked if bors can be configured to update the 'checks' section in a PR, but at least it comments with relevant URLs.

When CI is run you get a link to the gitlab pipeline with live logs via the little status indicator on the right (which is yellow as it runs).
Screenshot from 2020-04-09 11-44-33

**Comment by haampie** [comment]
Let's see if bors and gitlab work now

bors try

**Comment by haampie** [comment]
bors retry

**Comment by bcumming** [comment]
bors try

**Comment by bcumming** [comment]
bors try-

**Comment by bcumming** [comment]
bors try

**Comment by haampie** [comment]
Since gitlab was not configured yet, let me retry it

bors try-

**Comment by haampie** [comment]
bors try

**Comment by haampie** [comment]
Alright, I'll rest my case :sweat_smile:

**Comment by haampie** [comment]
@bcumming in principle this is good to go now.

I can create a new PR later that improves the performance a bit by reducing the number of resource allocations in slurm from 4 to 1.

**Comment by haampie** [comment]
In fact fixing the number of allocs seems easier than I thought. Let's see if it works

bors try

**Comment by haampie** [comment]
Now it should work!

bors try

**Comment by haampie** [comment]
bors try

**Comment by haampie** [comment]
@bcumming all useful features are implemented now :) ready for more review

**Comment by haampie** [comment]
bors try

**Comment by haampie** [comment]
bors try

**Pull request merged**